testapp/exam/code_server.py

```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
-        to prevent runaway code.
         Returns
         -------
 
         A tuple: (success, error message).
-
         """
-        if in_dir is not None and isdir(in_dir):
-            os.chdir(in_dir)
+        success = False
+        prev_handler = None
 
-        # Add a new signal handler for the execution of this code.
-        old_handler = signal.signal(signal.SIGALRM, timeout_handler)
-        signal.alarm(SERVER_TIMEOUT)
+        self._change_dir(self.in_dir)
+
+        if self.language == "python":
+            tb = None
+            test_code = self.create_test_case()
+            # Add a new signal handler for the execution of this code.
+            prev_handler = create_delete_signal_handler("new")
+
+            try:
+                submitted = compile(self.user_answer, '<string>', mode='exec')
+                g = {}
+                exec submitted in g
+                _tests = compile(test_code, '<string>', mode='exec')
+                exec _tests in g
+            except TimeoutException:
+                err = self.timeout_msg
+            except AssertionError:
+                type, value, tb = sys.exc_info()
+                info = traceback.extract_tb(tb)
+                fname, lineno, func, text = info[-1]
+                text = str(test_code).splitlines()[lineno-1]
+                err = "{0} {1} in: {2}".format(type.__name__, str(value), text)
+            except:
+                type, value = sys.exc_info()[:2]
+                err = "Error: {0}".format(repr(value))
+            else:
+                success = True
+                err = 'Correct answer'
+            finally:
+                del tb
+                # Set back any original signal handler.
+                create_delete_signal_handler("original", prev_handler)
+
+            # Cancel the signal
+            create_delete_signal_handler("delete")
 
-        success = False
-        tb = None
-        try:
-            submitted = compile(answer, '<string>', mode='exec')
-            g = {}
-            exec submitted in g
-            _tests = compile(test_code, '<string>', mode='exec')
-            exec _tests in g
-        except TimeoutException:
-            err = self.timeout_msg
-        except AssertionError:
-            type, value, tb = sys.exc_info()
-            info = traceback.extract_tb(tb)
-            fname, lineno, func, text = info[-1]
-            text = str(test_code).splitlines()[lineno-1]
-            err = "{0} {1} in: {2}".format(type.__name__, str(value), text)
-        except:
-            type, value = sys.exc_info()[:2]
-            err = "Error: {0}".format(repr(value))
         else:
-            success = True
-            err = 'Correct answer'
-        finally:
-            del tb
-            # Set back any original signal handler.
-            signal.signal(signal.SIGALRM, old_handler)
+            user_answer_file = {'C': 'submit.c', 'java': 'Test.java', 'scilab': 'function.sci',
```

Comment:
Use subclasses to segregate the different languages.  A method more than a screenful should be ruthlessly refactored.